pkg/endpoint/policy.go

```diff
@@ -163,81 +148,151 @@ func (e *Endpoint) setNextPolicyRevision(revision uint64) {
+//
+// buildMutex MUST be held, and not released until setDesiredPolicy and
+// updateRealizedState have been called
+//
+// There are a few fields that depend on this exact configuration of locking:
+//   - ep.desiredPolicy: ep.mutex must be locked between writing this and committing to
+//     the policy maps, or else policy drops may occur
+//   - ep.policyRevision: ep.mutex and ep.buildMutex must be held to write to this, or else
+//     the policyRevision may go backwards from no-op bumps from policy ingestion
+//   - ep.selectorPolicy: this may be nulled if the endpoints identity changes; we must
+//     check for this when committing. ep.mutex must be held
+//   - ep.realizedRedirects: this is read by external callers as part of policy generation,
+//     so ep.mutex must not be required to read this. Instead, both ep.mutex and ep.buildMutex
+//     must be held to write to this (i.e. we are deep in regeneration)
+//
+// Returns a result that should be passed to setDesiredPolicy after the endpoint's
+// write lock has been acquired, or err if recomputing policy failed.
+func (e *Endpoint) regeneratePolicy() (*policyGenerateResult, error) {
+	var err error
+
+	// lock the endpoint, read our values, then unlock
+	err = e.rlockAlive()
+	if err != nil {
+		return nil, err
+	}
 
 	// No point in calculating policy if endpoint does not have an identity yet.
 	if e.SecurityIdentity == nil {
 		e.getLogger().Warn("Endpoint lacks identity, skipping policy calculation")
-		return nil
+		e.runlock()
+		return nil, nil
+	}
+
+	// Copy out some values we care about, then unlock
+	forcePolicyCompute := e.forcePolicyCompute
+	securityIdentity := e.SecurityIdentity
+
+	result := &policyGenerateResult{
+		selectorPolicy:   e.selectorPolicy,
+		endpointPolicy:   e.desiredPolicy,
+		identityRevision: e.identityRevision,
 	}
+	e.runlock()
 
 	e.getLogger().Debug("Starting policy recalculation...")
 	stats := &policyRegenerationStatistics{}
 	stats.totalTime.Start()
+	defer func() {
+		stats.totalTime.End(err == nil)
+		e.updatePolicyRegenerationStatistics(stats, forcePolicyCompute, err)
+	}()
 
 	stats.waitingForPolicyRepository.Start()
 	repo := e.policyGetter.GetPolicyRepository()
 	repo.Mutex.RLock()
-	revision := repo.GetRevision()
-	defer repo.Mutex.RUnlock()
+	result.policyRevision = repo.GetRevision()
 	stats.waitingForPolicyRepository.End(true)
 
 	// Recompute policy for this endpoint only if not already done for this revision.
-	if !e.forcePolicyCompute && e.nextPolicyRevision >= revision {
+	if !forcePolicyCompute && e.nextPolicyRevision >= result.policyRevision && e.desiredPolicy != nil {
 		e.getLogger().WithFields(logrus.Fields{
 			"policyRevision.next": e.nextPolicyRevision,
-			"policyRevision.repo": revision,
+			"policyRevision.repo": result.policyRevision,
 			"policyChanged":       e.nextPolicyRevision > e.policyRevision,
 		}).Debug("Skipping unnecessary endpoint policy recalculation")
-
-		return nil
+		repo.Mutex.RUnlock()
```

Comment:
nit: I can't help but wonder if it'd be worth restructuring some of the function here a little in order to make use of defers for the repo mutex unlock. The code is correct, but sometimes developers can add a new error condition and neglect to properly handle locking when the lock release is spread throughout the function. Nothing a good round of review can't iron out, but from memory we have previously shipped error paths that don't unlock. It's usually not pretty, even if relatively easily found with Go tooling ;-) .

Reply:
If we can't leave a few landmines for future developers, what's even the fun?

Reply:
In all seriousness, I'll think about it for a bit. We did previously hold on to `repo.Mutex` for way too long

Reply:
Besides adding yet another function, I don't see an obvious fix. It would be good to drop `repo.Mutex` before `policy.Consume()` as, distillation is expensive.

Reply:
Yeah no worries. Again, this one's a nit, so up to your judgement. I'm fine with this as is, we can always iterate further on the code in subsequent PRs.